File: src/main/java/javabot/operations/UrlOperation.java
package javabot.operations;

import java.util.List;
import java.util.ArrayList;
import java.net.URLEncoder;
import java.nio.charset.Charset;
import java.io.UnsupportedEncodingException;

import javabot.Javabot;
import javabot.Message;
import javabot.BotEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Created Dec 16, 2008
 *
 * @author <a href="mailto:jlee@antwerkz.com">Justin Lee</a>
 */
public abstract class UrlOperation extends BotOperation {
    private static final Logger log = LoggerFactory.getLogger(UrlOperation.class);
    
    public UrlOperation(Javabot javabot) {
        super(javabot);
    }

    @Override
    public final List<Message> handleMessage(BotEvent event) {
        List<Message> messages = new ArrayList<Message>();
        String message = event.getMessage();
        String channel = event.getChannel();
        if (!message.startsWith(getTrigger())) {
            return messages;
        }
        message = message.substring(getTrigger().length());
        try {
            messages.add(new Message(channel, event,
                getBaseUrl() + URLEncoder.encode(message, Charset.defaultCharset().displayName())));
        } catch (UnsupportedEncodingException e) {
            log.error(e.getMessage(), e);
        }
        return messages;
    }

    protected abstract String getBaseUrl();

    protected abstract String getTrigger();

    @Override
    public List<Message> handleChannelMessage(BotEvent event) {
        return new ArrayList<Message>();
    }
}

File: src/main/kotlin/javabot/operations/UrlOperation.kt
package javabot.operations

import com.google.inject.Inject
import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.nio.charset.Charset
import javabot.Javabot
import javabot.Message
import javabot.dao.AdminDao
import org.slf4j.LoggerFactory

abstract class UrlOperation @Inject constructor(bot: Javabot, adminDao: AdminDao) :
    BotOperation(bot, adminDao) {

    override fun handleMessage(event: Message): List<Message> {
        val responses = arrayListOf<Message>()
        var message = event.value
        if (message.startsWith(getTrigger())) {
            message = message.substring(getTrigger().length)
            try {
                responses.add(
                    Message(
                        event,
                        getBaseUrl() +
                            URLEncoder.encode(message, Charset.defaultCharset().displayName()),
                    )
                )
            } catch (e: UnsupportedEncodingException) {
                log.error(e.message, e)
            }
        }
        return responses
    }

    protected abstract fun getBaseUrl(): String

    protected abstract fun getTrigger(): String

    companion object {
        private val log = LoggerFactory.getLogger(UrlOperation::class.java)
    }
}
